File: python/models/cnn_detector.py
"""
============================================================================
端到端CNN检测器模块
============================================================================
提供基于卷积神经网络的端到端MIMO检测器实现

主要组件:
  - CNNDetector: 端到端CNN检测器（改进版，使用匹配滤波初始化）
============================================================================
"""

import numpy as np
import torch
import torch.nn as nn
from typing import Optional


class CNNDetector(nn.Module):
    """
    改进版端到端CNN检测器
    
    改进点:
    1. 使用匹配滤波 (H^H * y) 作为初始估计输入
    2. 网络学习残差修正
    3. 移除过度的BatchNorm避免输出被压缩
    4. 添加输出缩放确保幅度正确
    
    Attributes:
        N: 符号维度
        hidden_channels: 隐藏层通道数
    """
    
    def __init__(self, N: int, hidden_channels: int = 64):
        """
        初始化CNN检测器
        
        Args:
            N: 符号维度 (信道矩阵大小 N×N)
            hidden_channels: 隐藏层通道数 (default: 64)
        """
        super().__init__()
        self.N = N
        self.hidden_channels = hidden_channels
        
        # 残差修正网络
        # 输入: 匹配滤波输出 [B, 2, N] (实部和虚部)
        self.refine_net = nn.Sequential(
            nn.Conv1d(2, hidden_channels, kernel_size=5, padding=2),
            nn.LeakyReLU(0.1),
            nn.Conv1d(hidden_channels, hidden_channels, kernel_size=5, padding=2),
            nn.LeakyReLU(0.1),
            nn.Conv1d(hidden_channels, hidden_channels, kernel_size=3, padding=1),
            nn.LeakyReLU(0.1),
            nn.Conv1d(hidden_channels, 2, kernel_size=3, padding=1),
        )
        
        # 信道特征提取 (用于调制残差)
        self.channel_encoder = nn.Sequential(
            nn.Conv2d(2, 32, kernel_size=3, padding=1),
            nn.LeakyReLU(0.1),
            nn.Conv2d(32, 32, kernel_size=3, padding=1),
            nn.LeakyReLU(0.1),
            nn.AdaptiveAvgPool2d((1, 1)),  # [B, 32, 1, 1]
        )
        
        # 缩放因子（可学习）
        self.scale = nn.Parameter(torch.ones(1))
        
        # 打印模型信息
        total_params = sum(p.numel() for p in self.parameters())
        print(f"CNNDetector: N={N}, hidden_channels={hidden_channels}, 参数量≈{total_params/1e3:.1f}K")
    
    def forward(self, y: torch.Tensor, H: torch.Tensor) -> torch.Tensor:
        """
        端到端CNN检测
        
        Args:
            y: 接收信号 [B, N, 1], complex64
            H: 信道矩阵 [B, N, N], complex64
        
        Returns:
            x_hat: 估计符号 [B, N, 1], complex64
        """
        B, N, _ = y.shape
        device = y.device
        
        # 1. 匹配滤波初始化: x_mf = H^H * y
        H_H = H.conj().transpose(1, 2)  # [B, N, N]
        x_mf = torch.matmul(H_H, y)  # [B, N, 1]
        
        # 归一化匹配滤波输出（避免数值问题）
        x_mf_abs_max = torch.abs(x_mf).reshape(B, -1).max(dim=1, keepdim=True)[0].unsqueeze(-1) + 1e-8
        x_mf_norm = x_mf / x_mf_abs_max
        
        # 2. 转换为实数格式 [B, 2, N]
        x_mf_squeezed = x_mf_norm.squeeze(-1)  # [B, N]
        x_real = torch.stack([x_mf_squeezed.real, x_mf_squeezed.imag], dim=1).float()  # [B, 2, N]
        
        # 3. 残差修正
        residual = self.refine_net(x_real)  # [B, 2, N]
        
        # 4. 残差连接
        x_refined = x_real + residual  # [B, 2, N]
        
        # 5. 转回复数并缩放到QPSK幅度
        x_hat = x_refined[:, 0, :] + 1j * x_refined[:, 1, :]  # [B, N]
        
        # 归一化到单位幅度，然后缩放到QPSK幅度 (1/sqrt(2))
        x_hat_normalized = x_hat / (torch.abs(x_hat) + 1e-8)
        x_hat_scaled = x_hat_normalized * (self.scale.abs() * 0.7071)
        
        x_hat_scaled = x_hat_scaled.unsqueeze(-1)  # [B, N, 1]
        
        return x_hat_scaled.to(torch.complex64)


if __name__ == "__main__":
    # 简单测试
    print("测试 CNNDetector:")
    
    N = 256
    B = 4
    
    model = CNNDetector(N=N, hidden_channels=64)
    
    # 生成测试数据
    y = torch.randn(B, N, 1, dtype=torch.complex64)
    H = torch.randn(B, N, N, dtype=torch.complex64)
    
    x_hat = model(y, H)
    print(f"输入 y 形状: {y.shape}")
    print(f"输入 H 形状: {H.shape}")
    print(f"输出 x_hat 形状: {x_hat.shape}")
    print(f"输出 dtype: {x_hat.dtype}")
    print(f"输出幅度均值: {torch.abs(x_hat).mean():.4f}")
    
    # 测试 CUDA 兼容性
    if torch.cuda.is_available():
        print("\n测试 CUDA 兼容性:")
        model_cuda = model.cuda()
        y_cuda = y.cuda()
        H_cuda = H.cuda()
        x_hat_cuda = model_cuda(y_cuda, H_cuda)
        print(f"CUDA 输出形状: {x_hat_cuda.shape}")
        print(f"CUDA 输出设备: {x_hat_cuda.device}")

File: matlab/matlab/detectors/cnn_detector.m
function x_hat = cnn_detector(y, H, params)
% =========================================================================
% CNN检测器 - 使用Python训练的参数
% =========================================================================
% 输入:
%   y         - 接收信号 [N × 1]
%   H         - 信道矩阵 [N × N]
%   params    - 从Python导出的参数结构体
%               包含CNN各层的权重和偏置
%
% 输出:
%   x_hat     - 估计符号 [N × 1]
%
% 算法:
%   1. 匹配滤波初始化: x_mf = H^H · y
%   2. 残差CNN修正
%   3. QPSK幅度缩放
% =========================================================================

    N = size(H, 1);
    
    % 1. 匹配滤波初始化
    H_H = H';
    x_mf = H_H * y;
    
    % 归一化
    x_mf_abs_max = max(abs(x_mf)) + 1e-8;
    x_mf_norm = x_mf / x_mf_abs_max;
    
    % 2. 转换为实数格式 [2, N] (实部和虚部)
    x_real = [real(x_mf_norm).'; imag(x_mf_norm).'];  % [2, N]
    
    % 3. 残差CNN修正
    % 注意: 这里需要实现1D卷积，MATLAB中需要手动实现
    residual = apply_refine_net(x_real, params);
    
    % 4. 残差连接
    x_refined = x_real + residual;
    
    % 5. 转回复数
    x_hat_raw = x_refined(1, :).' + 1j * x_refined(2, :).';
    
    % 6. 归一化到单位幅度，然后缩放到QPSK幅度
    x_hat_normalized = x_hat_raw ./ (abs(x_hat_raw) + 1e-8);
    scale = abs(params.scale) * 0.7071;
    x_hat = x_hat_normalized * scale;
end


function out = apply_refine_net(x, params)
% 应用残差修正网络
% 输入: x [2, N]
% 输出: out [2, N]

    % 简化实现：如果没有完整的CNN参数，使用恒等映射
    if ~isfield(params, 'refine_net_0_weight')
        out = zeros(size(x));
        return;
    end
    
    % Conv1d层实现
    % Layer 0: Conv1d(2, hidden, kernel=5, padding=2)
    h = conv1d_layer(x, params.refine_net_0_weight, params.refine_net_0_bias, 2);
    h = leaky_relu(h, 0.1);
    
    % Layer 2: Conv1d(hidden, hidden, kernel=5, padding=2)
    h = conv1d_layer(h, params.refine_net_2_weight, params.refine_net_2_bias, 2);
    h = leaky_relu(h, 0.1);
    
    % Layer 4: Conv1d(hidden, hidden, kernel=3, padding=1)
    h = conv1d_layer(h, params.refine_net_4_weight, params.refine_net_4_bias, 1);
    h = leaky_relu(h, 0.1);
    
    % Layer 6: Conv1d(hidden, 2, kernel=3, padding=1)
    out = conv1d_layer(h, params.refine_net_6_weight, params.refine_net_6_bias, 1);
end


function out = conv1d_layer(x, weight, bias, padding)
% 1D卷积层
% 输入: x [C_in, N]
% 权重: weight [C_out, C_in, kernel_size]
% 输出: out [C_out, N]

    [C_in, N] = size(x);
    [C_out, ~, K] = size(weight);
    
    % 填充
    x_padded = [zeros(C_in, padding), x, zeros(C_in, padding)];
    
    % 卷积
    out = zeros(C_out, N);
    for c_out = 1:C_out
        for c_in = 1:C_in
            kernel = squeeze(weight(c_out, c_in, :));
            out(c_out, :) = out(c_out, :) + conv(x_padded(c_in, :), flip(kernel), 'valid');
        end
        out(c_out, :) = out(c_out, :) + bias(c_out);
    end
end


function out = leaky_relu(x, alpha)
% Leaky ReLU激活函数
    out = max(x, alpha * x);
end
